File: Java/labs/lab05_Prototype/task_2_2/src/Engine.java
public class Engine implements CloneableEngine {

    @Override
    public Engine clone() {
        return new Engine(this);
    }

    public enum Fuel {
        Petrol,
        Diesel,
        Electric
    }

    final int power;
    final int torque;
    final Fuel fuel;
    final float volume;

    public Engine(int power, int torque, Fuel fuel, float volume) {
        this.power = power;
        this.torque = torque;
        this.fuel = fuel;
        this.volume = volume;
    }

    public Engine(Engine engine) {
        this.power = engine.power;
        this.torque = engine.torque;
        this.fuel = engine.fuel;
        this.volume = engine.volume;
    }

    @Override
    public String toString() {
        return "Engine{" +
                "power=" + power +
                ", torque=" + torque +
                ", fuel=" + fuel +
                ", volume=" + volume +
                '}';
    }
}

File: Cpp/lab05_Prototype/task_2_2/src/engine.cpp
#include "engine.hpp"


Engine::Engine(int power, int torque, Engine::Fuel fuel, float volume)
: power(power), torque(torque), fuel(fuel), volume(volume) {}

Engine::Engine(Engine const& other)
: power(other.power), torque(other.torque),
fuel(other.fuel), volume(other.volume) {}

Engine* Engine::clone() const
{
  return new Engine(*this);
}

std::ostream& operator<<(std::ostream& os, Engine const& engine)
{
  os << "Engine{power=" << engine.power
     << ", torque=" << engine.torque
     << ", fuel=";

  switch (engine.fuel)
  {
    case Engine::Fuel::Petrol:
      os << "Petrol";
      break;
    case Engine::Fuel::Diesel:
      os << "Diesel";
      break;
    case Engine::Fuel::Electric:
      os << "Electric";
      break;
  }

  os << ", volume=" << engine.volume << "}";

  return os;
}
